src/os/filestore/FileStore.cc

```diff
@@ -1794,11 +1806,249 @@ int FileStore::umount()
     timer.shutdown();
   }
 
+#ifdef HAVE_LIBAIO
+  destroy_read_completion_threads();
+#endif
+
   // nothing
   return 0;
 }
 
+#ifdef HAVE_LIBAIO
+int FileStore::create_read_completion_threads()
+{
+  // NB: m_filestore_async_threads may be 0.
+  int error = 0;
+  dout(0) << "FileStore: creating " << m_filestore_async_threads
+	  << " read completion threads." << dendl;
+  for (int i = 0; i < m_filestore_async_threads; ++i) {
+    ReadCompletionThread *thr = new ReadCompletionThread(this);
+    /*
+     * FIXME: do we need to set priority?
+     * int r = thr->set_ioprio(thr->ioprio_class, thr->ioprio_priority);
+     */
+    read_completion_threads.push_back(thr);
+    error = thr->Start();
+    if (error)
+      break;
+  }
+
+  if (error) {
+    derr << "unable to create read completion threads, error = "
+	 << error << dendl;
+    for (auto rct : read_completion_threads)
+      delete rct;
+    read_completion_threads.clear();
+  }
+ 
+  return error;
+}
+
+void FileStore::destroy_read_completion_threads()
+{
+  if (read_completion_threads.size() == 0)
+    return;
+  dout(0) << "FileStore: destroying " << read_completion_threads.size()
+	  << " read completion threads." << dendl;
+
+  for (auto rct : read_completion_threads)
+    delete rct;
+
+  dout(0) << "FileStore:  destroy_read_completion_threads done" << dendl;
+  read_completion_threads.clear();
+}
+
+int FileStore::async_read_dispatch(
+  Sequencer *posr,
+  Context *ctx,
+  const coll_t& cid,
+  const ghobject_t& oid,
+  uint64_t off,
+  size_t len,
+  bufferlist *bl,
+  uint32_t op_flags,
+  Context *blessed_context)
+{
+  OpSequencer *osr = may_create_op_sequencer(posr);
+  // PGID-to-completion thread mapping:
+  int thread_index = osr->id % read_completion_threads.size();
+
+  FDRef fd;
+  off64_t offset = (off64_t) off;
+  int r = lfn_open(cid, oid, false, &fd, 0, true);
+  if (r < 0) {
+    dout(10) << "FileStore::async_read_dispatch(" << cid << "/" << oid << ") open error: "
+	     << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  if (len == 0) {
+    struct stat st;
+    memset(&st, 0, sizeof(struct stat));
+    int r = ::fstat(**fd, &st);
+    assert(r == 0);
+    len = st.st_size;
+  }
+
+#ifdef HAVE_POSIX_FADVISE
+  if (op_flags & CEPH_OSD_OP_FLAG_FADVISE_RANDOM)
+    posix_fadvise(**fd, offset, len, POSIX_FADV_RANDOM);
+  if (op_flags & CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL)
+    posix_fadvise(**fd, offset, len, POSIX_FADV_SEQUENTIAL);
+#endif
+
+  // Extend to 512 bytes since direct IO is being used
+  unsigned alignment = len % 512;
```

Comment:
Agreed. I will add some BLKBSZGET ioctl() processing to get the correct alignment.